pilot/pkg/status/state.go

```diff
@@ -0,0 +1,304 @@
+// Copyright 2019 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package status
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"k8s.io/apimachinery/pkg/runtime/schema"
+
+	"k8s.io/apimachinery/pkg/labels"
+
+	"github.com/ghodss/yaml"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/apis/meta/v1beta1"
+	"k8s.io/client-go/dynamic"
+	"k8s.io/client-go/informers"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/rest"
+	"k8s.io/utils/clock"
+
+	"istio.io/pkg/log"
+)
+
+var scope = log.RegisterScope("status",
+	"component for writing distribution status to istiio CRDs", 0)
+
+type Progress struct {
+	AckedInstances int
+	TotalInstances int
+}
+
+func (p *Progress) PlusEquals(p2 Progress) {
+	p.TotalInstances += p2.TotalInstances
+	p.AckedInstances += p2.AckedInstances
+}
+
+type DistributionController struct {
+	mu               sync.RWMutex
+	CurrentState     map[Resource]map[string]Progress
+	ObservationTime  map[string]time.Time
+	UpdateInterval   time.Duration
+	client           dynamic.Interface
+	clock            clock.Clock
+	knownResources   map[schema.GroupVersionResource]dynamic.NamespaceableResourceInterface
+	currentlyWriting ResourceLock
+	StaleInterval    time.Duration
+	QPS              float32
+	Burst            int
+}
+
+func (c *DistributionController) Start(restConfig *rest.Config, namespace string, stop <-chan struct{}) {
+	scope.Info("Starting status leader controller")
+	// default UpdateInterval
+	if c.UpdateInterval == 0 {
+		c.UpdateInterval = 200 * time.Millisecond
+	}
+	// default StaleInterval
+	if c.StaleInterval == 0 {
+		c.StaleInterval = time.Minute
+	}
+	if c.clock == nil {
+		c.clock = clock.RealClock{}
+	}
+	c.CurrentState = make(map[Resource]map[string]Progress)
+	c.ObservationTime = make(map[string]time.Time)
+	c.knownResources = make(map[schema.GroupVersionResource]dynamic.NamespaceableResourceInterface)
+	var err error
+	restConfig.QPS = c.QPS
+	restConfig.Burst = c.Burst
+	if c.client, err = dynamic.NewForConfig(restConfig); err != nil {
```

Comment:
Same - we have at least one dynamic already used in validation.

That's why we did Istiod refactoring - to stop the inconsistent duplication of boilerplate.

Reply:
This clientset needs to have configurable QPS and Burst for the unique constraints of status.  If there is a way to modify these after initialization, I'd be happy to use that.

Reply:
So you're relying on the client library to control the rate of write ? 
I would use a more explicit control, in the code that actually writes. I'm sure there is a library we use somewhere else in the code.

Reply:
Also: add comment and make it clear. I don't know if users expect that istiod will have 2 QPS and burst rates, we had problems and post-mortem with Istiod killing apiservers. 
I would feel much better if istiod as a whole had a clear max QPS - instead of various features each defining their own and user having to add them app and guess.

Reply:
Added explanatory comment for QPS and Burst.